bootstrap/cmd/bootstrap/app/kfctlServer.go

```diff
@@ -60,15 +301,214 @@ func (r *kfctlServer) RegisterEndpoints() {
 	http.Handle(KfctlCreatePath, optionsHandler(createHandler))
 }
 
-// Apply runs apply on a ksonnet application.
-func (r *kfctlServer) CreateDeployment(ctx context.Context, req CreateRequest) (*CreateResponse, error) {
-	// TODO(jlewi):
-	// 1. Do IAM check
-	// 2. Check if service exists by sending request
-	// 3. If service doesn't exist create a service and statefulset
-	log.Infof("Recieved deployment request: %+v", req)
-	return nil, &httpError{
-		Message: "CreateDeployment isn't implemented yet.",
-		Code:    http.StatusNotImplemented,
+// isMatch checks whether the incoming request is a match for the deployment
+// that is already started. If not it is rejected.
+func isMatch(current *kfdefsv2.KfDef, new *kfdefsv2.KfDef) bool {
+	// Ensure neither is nil
+	if current == nil {
+		return true
 	}
+
+	if new == nil {
+		return false
+	}
+
+	if current.Name == "" && current.Spec.Project == "" && current.Spec.Zone == "" {
+		return true
+	}
+
+	if current.Spec.Project != new.Spec.Project {
+		return false
+	}
+
+	if current.Spec.Zone != new.Spec.Zone {
+		return false
+	}
+
+	if current.Name != new.Name {
+		return false
+	}
+
+	return true
+}
+
+// prepareSecrets prepares the secrets in the request.
+// Literal secrets are converted to environment variables except for the GcpAccessToken which is removed.
+//
+// TODO(https://github.com/kubeflow/kubeflow/issues/3592) Once the apply methods take a context we should be able
+// to use that and not rely on the environment
+func prepareSecrets(d *kfdefsv2.KfDef) {
+	secrets := []kfdefsv2.Secret{}
+
+	for _, s := range d.Spec.Secrets {
+		// Don't pass along the access token
+		if s.Name == gcp.GcpAccessTokenName {
+			continue
+		}
+
+		if s.SecretSource.LiteralSource != nil {
+			n := "KFCTL_" + s.Name
+			s.SecretSource.EnvSource = &kfdefsv2.EnvSource{
+				Name: n,
+			}
+
+			os.Setenv(n, s.SecretSource.LiteralSource.Value)
+			s.SecretSource.LiteralSource = nil
+
+		}
+
+		secrets = append(secrets, s)
+	}
+
+	d.Spec.Secrets = secrets
+}
+
+// CreateDeployment creates the deployment.
+//
+// Not thread safe
```

Comment:
same here.  I think my concern is that is it possible that we have multiple invocations?  should we make server codes idempotent?